Spark 3.2

echeipesh opened the pull request:
Pushing for Spark 3.2
- [x] get tests green
- [x] update to spark-testing-base for more speed
- [x] get code review and/or cleanup the changes
- [x] figure out what to do with GeoTrellis layer data source
- [x] update python package
  - ~cleanup dependencies: disentangle the release wheel from pweave~
  - ~grok the python environment, consider introducing `poetry` instead of conda~
- [x] run some jobs in spark 3.2 environment

metasim committed:
GitHub actions build.

- Split out docs build into separate workflow.
- Removed umlimit call (not needed)

metasim committed:
bumped dev version

metasim committed:
bumped dev version

Spark 3.2

Lets get it compiled, spark 2 support is well out the window anyway.

metasim committed:
CI fix.

metasim committed:
Dependency updates.

metasim committed:
Spark 3.1.3

echeipesh committed:
withNewChildrenInternal

echeipesh committed:
Try Aggregator implemtnation

echeipesh committed:
more explicit

echeipesh commented on core/src/main/scala/org/locationtech/rasterframes/expressions/BinaryRasterFunction.scala:
```diff
@@ -25,13 +25,14 @@ import com.typesafe.scalalogging.Logger
 /** Operation combining two tiles or a tile and a scalar into a new tile. */
-trait BinaryRasterFunction extends BinaryExpression with RasterResult {
+trait BinaryRasterFunction extends BinaryExpression with RasterResult { self: HasBinaryExpressionCopy =>
+  override protected def withNewChildrenInternal(newLeft: Expression, newRight: Expression): Expression = copy(newLeft, newRight)
```
Tried being clever here with structural self type in order to call copy in the super-type and I really don't think this is worth the brain cells. Down this road is creating things like `RFBinaryExpression` that has this little bit of magic. However, it will not always work, for instance when the case class does not use standard `left`, `right` or `child` constructor names. So I'm probably going to revert this and do some more c/p.

echeipesh replied:
Yes, this was a bad idea

echeipesh commented on core/src/main/scala/org/locationtech/rasterframes/expressions/transformers/ReprojectGeometry.scala:
```diff
@@ -49,12 +49,12 @@ import org.locationtech.rasterframes.model.LazyCRS
   override def nodeName: String = "st_reproject"
+  def first: Expression = geometry
+  def second: Expression = srcCRS
+  def third: Expression = dstCRS
```
Couple of places the Expression case classes actually have meaningful names on the constructor args which feels nice. Where they don't the naming convention changed.

echeipesh committed:
Fix UDF style Aggregates

pomadchin commented:
I'm going to bump the min Spark version in GT as well https://github.com/locationtech/geotrellis/pull/3389 
Thogh it should not affect this PR much.

echeipesh commented:
@pomadchin It'd be nice to get 3.6.3 out with spark 3.2.0 and then 3.7.0 with CES3 and spark 3.2.1. That'd allow this to sync with frameless 0.11 and later with 0.12+.

pomadchin commented:
@echeipesh 🚀 https://github.com/locationtech/geotrellis/pull/3471 + I'm dropping Spark 3.0 over there https://github.com/typelevel/frameless/pull/637 (frameless will be published for 3.1.x, 3.2.x, and 3.3.x)

echeipesh committed:
Bring in the Kryo setup

GT settings bring in jackson classes and with shading it was getting weird

echeipesh committed:
Register functions directly

this is a starting point

echeipesh committed:
Bump versions

echeipesh committed:
Landsat PDS is gone :(

echeipesh committed:
Fix Resample and ResampleNearest

Also untangle the super weird inheritance relationship between the two

echeipesh committed:
fix masking functions

Made them more direct. Good for fixing things and better for performance because these versions don't need to create intermediate mask tiles.

echeipesh committed:
Fix test: 6900 bit at position 4 is 1 -- expect NODATA after mask

echeipesh committed:
TileRasterizerAggregate expects column in rf_raster_proj order

This is a change but it's towards less surprising

echeipesh committed:
Use spark-testing-base - core tests green

- fixed weird init order in tests
- all tests share same context now thanks to base
- exclude scala-xml from tests

echeipesh committed:
Update StacApiDataSourceTest.scala

echeipesh committed:
disable GeoTrellisDataSourceSpec

Who read these anyway?

echeipesh commented on datasource/src/test/scala/org/locationtech/rasterframes/datasource/geotrellis/GeoTrellisDataSourceSpec.scala:
```diff
@@ -48,7 +48,8 @@ import org.scalatest.{BeforeAndAfterAll, Inspectors}
 import scala.math.{max, min}
 
-class GeoTrellisDataSourceSpec extends TestEnvironment with BeforeAndAfterAll with Inspectors with DataSourceOptions {
+trait GeoTrellisDataSourceSpec extends TestEnvironment with BeforeAndAfterAll with Inspectors with DataSourceOptions {
```
Very little motivation to fix this. Note that the push-down predicates also got busted with move to spark 3.2. 

Can't imagine ever using this rather than reading COGs or even Zarr. Unless somebody would like to fix this I would like to suggest to remove this feature altogether to keep things tidy.

echeipesh replied:
@metasim @pomadchin What do you guys think?

pomadchin replied:
:+1:

echeipesh committed:
Shade caffeine

popular caching library

echeipesh commented on project/RFDependenciesPlugin.scala:
```diff
@@ -70,9 +71,9 @@ object RFDependenciesPlugin extends AutoPlugin {
+    rfSparkVersion := "3.2.1",
+    rfGeoTrellisVersion := "3.6.3",
+    rfGeoMesaVersion := "3.4.1"
+    //excludeDependencies += "log4j" % "log4j"
```
Wait, why was this excluded?

pomadchin replied:
I think it's another way to solve the CVE problem but idk!

echeipesh committed:
Merge branch 'develop' into spark-3.2

echeipesh committed:
boop

pomadchin commented on core/src/main/scala/org/locationtech/rasterframes/expressions/generators/RasterSourceToTiles.scala:
```diff
@@ -84,6 +84,8 @@ case class RasterSourceToTiles(children: Seq[Expression], bandIndexes: Seq[Int],
     }
   }
+
+  override protected def withNewChildrenInternal(newChildren: IndexedSeq[Expression]): Expression = copy(children=newChildren)
```
minor: seems the formatting is a bit off there `children=newChildren` => `children = newChildren`

echeipesh replied:
Heh, python formatting bleeding through.

pomadchin commented on core/src/main/scala/org/locationtech/rasterframes/expressions/package.scala:
```diff
@@ -60,114 +64,166 @@ package object expressions {
+      registry.registerFunction(id, info, builder)
+    }
+
+    def register1[T <: Expression : ClassTag](
+      name: String,
+      builder: Expression => T
+    ): Unit = registerFunction[T](name, None){ args => builder(args(0))
+    }
+
+    def register2[T <: Expression : ClassTag](
+      name: String,
+      builder: (Expression, Expression) => T
+    ): Unit = registerFunction[T](name, None){ args => builder(args(0), args(1)) }
+
+    def register3[T <: Expression : ClassTag](
+      name: String,
+      builder: (Expression, Expression, Expression) => T
+    ): Unit = registerFunction[T](name, None){ args => builder(args(0), args(1), args(2)) }
+
+    def register5[T <: Expression : ClassTag](
+      name: String,
+      builder: (Expression, Expression, Expression, Expression, Expression) => T
+    ): Unit = registerFunction[T](name, None){ args => builder(args(0), args(1), args(2), args(3), args(4)) }
```
So far the only ugly place!

echeipesh replied:
Thank you for fixing it!

pomadchin reviewed (approved):
LGTM!

pomadchin commented on core/src/main/scala/org/locationtech/rasterframes/expressions/package.scala:
```diff
@@ -46,6 +46,10 @@ import scala.reflect.runtime.universe._
 package object expressions {
+  type HasTernaryExpressionCopy = {def copy(first: Expression, second: Expression, third: Expression): Expression}
+  type HasBinaryExpressionCopy = {def copy(left: Expression, right: Expression): Expression}
+  type HasUnaryExpressionCopy = {def copy(child: Expression): Expression}
```
Those can be expensive I think but mb it's fine 🤔

echeipesh replied:
What do you mean by expensive here?

pomadchin replied:
Structural types in Scala2 rely on runtime reflection; that's not a big deal but it just got caught by my eye.

pomadchin replied:
A bit better practice is usually to avoid them by having named traits 💭

echeipesh replied:
Thank you for calling this out. Can't have named traits because it's trying to call a copy method that will be added by the case class. So it HAS to be runtime reflection. But the whole point of this API change was that runtime reflection was killing Catalyst performance. So while its ugly its better to just go ahead and implement the `withNewChildrenInternal` methods in the case class definitions.

pomadchin committed:
Expressions constructors toSeq conversion

pomadchin commented:
CI is dead? https://app.circleci.com/pipelines/github/locationtech/rasterframes?utm_campaign=workflow-failed&utm_medium=email&utm_source=notification

echeipesh committed:
Downgrade scaffeine to 4.1.0 for JDK 8 support in caffeine 2.9

echeipesh committed:
pyspark version 3.2.1

echeipesh committed:
why exclude log4j ? tests need it

echeipesh committed:
Fix formatting

pomadchin committed:
Fix Expressions arity issue

pomadchin committed:
Add .jvmopts

echeipesh committed:
Fix: Mask operations preserver the target tile cell type

pomadchin commented on core/src/main/scala/org/locationtech/rasterframes/expressions/transformers/MaskExpression.scala:
```diff
@@ -0,0 +1,74 @@
+
+import spire.syntax.cfor._
+
+trait MaskExpression { self: Expression =>
```
We can cut a fresh GT release with the fixed mask operation! Sadly due to the CircleCI breach the most up to date snapshot has not been released.

echeipesh replied:
Actually this is kind of fine. The logic here is slightly different from the standard implementation because we know we're mapping over int values. I would bet that it would be more helpful in the future to have full control over this than saving a few lines of code.

echeipesh committed:
Pin GitHub Actions to ubuntu-20.04

echeipesh committed:
Implement withNewChildrenInternal directly

avoid reflection which is done at runtime by structural types

echeipesh committed:
Remove python build from CI

It needs more work at another time

echeipesh commented:
I've been looking at the python side of the project enough to form strong enough opinions of it. I think it's actually going to need some work and it doesn't make sense to do it in this PR.

- make the build PEP 517 compatible
- ditch use of `conda` for `poetry` (we're not building a conda package here anyway)
- separate the the `pweave` deps from the release deps
- decide if shipping jars with `.whl` still makes sense
  - it was a nice attempt at nicety but unless you start the process from `ipython` instead of `pyspark` it is just dead weight.
- `pyrasterframes` python files are in the sbt project structure
  - Kind of the only benefit of this is to couple the production of assembly jar with the building of the .whl
  - Not sure the weirdness of this justifies the setup

So that's a lot of moving stuff around potentially and it would be much cleaner to make that it's own PR.
